direct/core/src/main/java/cz/o2/proxima/direct/view/TimeBoundedVersionedCache.java

```diff
@@ -129,6 +133,36 @@ synchronized void scan(
     }
   }
 
+  /** Clear records that are older than the given timestamp. */
+  public synchronized void clearStaleRecords(long cleanTime) {
+    List<Entry<String, NavigableMap<String, NavigableMap<Long, Payload>>>> candidates =
+        cache
+            .entrySet()
+            .stream()
+            .filter(
+                e -> e.getValue().values().stream().anyMatch(v -> v.floorEntry(cleanTime) != null))
+            .collect(Collectors.toList());
+    for (Entry<String, NavigableMap<String, NavigableMap<Long, Payload>>> candidate : candidates) {
+      candidate
+          .getValue()
+          .forEach(
+              (key, value) -> {
+                List<Long> toRemove = new ArrayList<>(value.headMap(cleanTime).keySet());
+                toRemove.forEach(value::remove);
+              });
+      candidate
+          .getValue()
+          .entrySet()
+          .stream()
+          .filter(e -> e.getValue().isEmpty())
+          .collect(Collectors.toList())
+          .forEach(e -> candidate.getValue().remove(e.getKey()));
+      if (candidate.getValue().isEmpty()) {
+        cache.remove(candidate.getKey());
+      }
+    }
+  }
```

Comment:
It would be nice to simplify this. It actually took me a while to understand what exactly is the method doing.

Maybe something like?
```suggestion
    final List<String> emptyKeys = new ArrayList<>();
    cache.forEach(
        (key, attributes) -> {
          final List<String> emptyAttributes = new ArrayList<>();
          attributes.forEach(
              (attribute, entries) -> {
                // Remove entries with timestamp lower than `cleanTime`.
                Iterables.removeIf(entries.headMap(cleanTime, false).keySet(), timestamp -> true);
                if (entries.isEmpty()) {
                  emptyAttributes.add(attribute);
                }
              });
          emptyAttributes.forEach(attributes::remove);
          if (attributes.isEmpty()) {
            emptyKeys.add(key);
          }
        });
    emptyKeys.forEach(cache::remove);
```

Reply:
simplified. It surprises me, that it is possible to call `removeIf` on `keySet`. But apparently it works.

Reply:
It's supported to call .remove() on keySet and entrySet iterators. It's a common technique to avoid ConcurrentModificationException.